flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/optimization/GradientDescent.scala

```diff
@@ -192,10 +190,18 @@ abstract class GradientDescent extends IterativeSolver {
       (left, right) =>
         val (leftGradVector, leftCount) = left
         val (rightGradVector, rightCount) = right
-        // Add the left gradient to the right one
-        BLAS.axpy(1.0, leftGradVector.weights, rightGradVector.weights)
+
+        // make the left gradient dense so that the following reduce operations (left fold) reuse
+        // it. This strongly depends on the underlying implementation of the ReduceDriver
```

Comment:
The reduce is implemented somewhat like a left fold operation, just without having an initial value. Thus you should not convert the right element, because it will always be the newly read input element.